solutions/apio-10-commando.mdx

```diff
@@ -0,0 +1,115 @@
+---
+id: apio-10-Commando
+source: APIO 2010
+title: Commando
+author: Dustin Miao
+---
+
+# Explanation
+
+Let $\texttt{dp}[i]$ denote the maximum effectiveness of the prefix of length $i$ of the unit. The transitions are defined as $$\texttt{dp}[i] = \max_{1 \leq j < i} \{\texttt{dp}[j] + ax^2+bx+c\}$$, where $x = \texttt{sum}(j + 1, i) = \sum_{k = j + 1}^i x_k$. Define a prefix sum array $\texttt{pre} = \sum_{k = 1}^i x_k$. We can then express $x$ as $\texttt{pre}(i) - \texttt{pre}(j)$. 
+
+Evaluate the first equation with our new value of $x = \texttt{pre}(i) - \texttt{pre}(j)$
+- $\texttt{dp}(i) = \max_{1 \leq j < i} \{\texttt{dp}[j] + ax^2+bx+c\}$
+- $\texttt{dp}(i) = \max_{1 \leq j < i} \{\texttt{dp}[j] + a[\texttt{pre}(i) - \texttt{pre}(j)]^2+b[\texttt{pre}(i) - \texttt{pre}(j)]+c\}$
+- $\texttt{dp}(i) = \max_{1 \leq j < i} \{\texttt{dp}[j] + [a\cdot\texttt{pre}(i)^2 - 2a\cdot\texttt{pre}(i)\texttt{pre}(j) + a\cdot\texttt{pre}(j)^2] + b\cdot\texttt{pre}(i)-b\cdot\texttt{pre}(j)+c\}$
+
+We will factor terms that do not depend on $j$ outside of the $\max$ statement.
+
+$$\texttt{dp}(i) = \max_{1 \leq j < i} \{\texttt{dp}[j] - 2a\cdot\texttt{pre}(i)\texttt{pre}(j) + a\cdot\texttt{pre}(j)^2 - b\cdot\texttt{pre}(j)\} + a\cdot\texttt{pre}(i)^2 + b\cdot\texttt{pre}(i) + c$$
+
+We will group terms that depend only on $j$ with $\texttt{dp}(j)$.
+
+$$\texttt{dp}(i) = \max_{1 \leq j < i} \{(\texttt{dp}[j] - a\cdot\texttt{pre}(j)^2 - b\cdot\texttt{pre}(j)) - 2a\cdot\texttt{pre}(i)\texttt{pre}(j)\} + a\cdot\texttt{pre}(i)^2 + b\cdot\texttt{pre}(i) + c$$
+
+To efficiently handle transitions in $\mathcal{O}(\log n)$ or $\mathcal{O}(1)$, we will use the convex hull optimization.
+
+**Time Complexity:** $\mathcal{O}(n \log n)$
+
+# Implementation 
+
+<LanguageSection>
+
+<CPPSection>
+  
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+using ll = long long;
+
+inline namespace _LineContainer {
+    bool _Line_Comp_State; 
+    struct Line {
+        mutable long long k, m, p;
+        bool operator<(const Line& o) const {
+            return _Line_Comp_State ? p < o.p : k < o.k;
+        }
+    };
+    
+    struct LineContainer : multiset<Line> {
+        const long long inf = LLONG_MAX;
+ 
+        inline long long div(long long a, long long b) { return a / b - ((a ^ b) < 0 && a % b); }
+ 
+        bool isect(iterator x, iterator y) {
+            if (y == end()) { 
+                x->p = inf; 
+                return false; 
+            }
+            if (x->k == y->k) 
+                x->p = x->m > y->m ? inf : -inf;
+            else 
+                x->p = div(y->m - x->m, x->k - y->k);
+            return x->p >= y->p;
+        }
+
+        /**
+         * @param k Slope
+         * @param m Y-Intercept
```

Comment:
1. why not just use the names `slope` and `y_inter` instead of putting it in the documents
2. actually don't think a comment explaining this function is necessary but it'd be nice